### ISSUE TITLE
VM name prefix not matching inventory_hostname causes vars not to be loaded

### ISSUE THREAD

**Issue by tumbl3w33d** [ansible-community/molecule-hetznercloud#89]
Hello,

TL;DR:

- molecule knows instances by their name defined in the `platforms` section
- `molecule-hetznercloud` prefixes instances (by default a dynamic string, computed by md5 of the scenario path)
- if you use an additional inventory that works with the hcloud API, you must be aware the instances names in hcloud won't match the host names in your play, i.e. host_vars also won't match
- my workaround is to define a static prefix (`RESOURCE_NAMESPACE`) and remove that prefix on-the-fly from the instance names in my dynamic inventory

**Long story:**

I define my instances in the `platforms` attribute like this:
```yaml
platforms:
  - name: "${$USER}-app-01"
    groups:
      - my_servers
 […]
  ```
and in my `converge.yml` I target the hosts like
```yaml
- name: Converge
  hosts: my_servers
```
This basically works. However, I also link my dynamic inventory
```yaml
provisioner:
[…]
  inventory:
    links:
      hosts: "../../../../inventories/hcloud_inventory.py"
```
which picks up my hosts and sets some hostvars each. Now the problem is, my tests run against the group's hosts like `tumble-app-01`, but in the background the hcloud driver prefixes VMs with a resource namespace (with fallback to a md5 of the scenario path which effectively is the value, because I didn't set that env myself).

The result is that my dynamic inventory knows a host like `<md5>-tumble-app-01` which is not the same as the one used in the molecule run (`tumble-app-01`) which in turn means that my hostvars are not loaded.

I added a workaround to my dynamic inventory to remove that prefix from the host, when a certain env var is set:

```python
    # for molecule-based tests running in hcloud
    if "VM_PREFIX_TO_REMOVE" in os.environ:
        vm_name = vm_name.replace(os.environ["VM_PREFIX_TO_REMOVE"], "", 1)
 ```
 That's not pretty, but I could live with it. However, I struggle to dynamically set that env for the run:
 
 ```yaml
provisioner:
[…]
  env:
    ANSIBLE_ROLES_PATH: "../../../../roles:../../../../common-roles"
    INV_HCLOUD_TOKEN: "${HCLOUD_TOKEN}"
    #VM_PREFIX_TO_REMOVE: "{{ '${MOLECULE_SCENARIO_DIRECTORY}' | md5 }}-" # this does not work, no interpolation
    VM_PREFIX_TO_REMOVE: 19fa9d8fc0b584b2b5e2704a69929e2d-
```
The statically set prefix leads to my inv removing that part of the host name and my play picks up the vars as expected. What I can do is setting `RESOURCE_NAMESPACE` there, so I know the static string that I need to remove from the host names.

It's all a little awkward. I guess there's no way for you to influence the host names molecule works with. I tried targeting the hosts directly instead of the group like
```yaml
  hosts:
    - "{{ lookup('env', 'MOLECULE_SCENARIO_DIRECTORY') | md5 }}-app-01"
    - "{{ lookup('env', 'MOLECULE_SCENARIO_DIRECTORY') | md5 }}-app-02"
```
and that basically works, however, it's somehow out of sync with the molecule universe so it doesn't know how to reach these hosts and the play fails with unreachable (permission denied).

I guess I will continue with the static prefix for now as that does the trick for me due to the workaround in my inventory. Feel free to close the issue if you think that's the best we can do.

**Comment by jooola** [comment]
I am not sure to fully understand your problem. Could you please explain your goal? 

- Are you trying to reuse existing servers with molecule?
- Are you trying to mix the inventory plugin with molecule? What is the purpose of this?
- Are you saying the molecule servers are polluting your inventory?

Here are a few tips:

Are you required to use the same project for both your molecule tests and the rest of your infrastructure? Maybe you should use a separate project for molecule?

You can filter your inventory hosts using label selectors. All the resources that molecule creates will have a label `molecule={{ resource_namespace }}`, so you could filter out all the molecule host using `label_selector: "!molecule"` in your inventory configuration.

In the inventory, you can automatically add a hostname to your hosts, See the [`hostname`](https://docs.ansible.com/ansible/devel/collections/hetzner/hcloud/hcloud_inventory.html#parameter-hostname) options.

**Comment by tumbl3w33d** [comment]
> Are you trying to reuse existing servers with molecule?

No, not really (at molecule test start time they do not exist, but once they have been created by molecule, I need some metadata from hcloud API), but you will understand it better when you read my answer to "what is the purpose of this" below.

> Are you trying to mix the inventory plugin with molecule?

Not sure we have a common understanding of "inventory plugin". I'm using a dynamic inventory script (not the official one you linked, mine is a lot more complex) in addition to molecule's platforms attribute. The [molecule documentation](https://ansible.readthedocs.io/projects/molecule/configuration/#molecule.provisioner.ansible.Ansible) describes that:

> Like above, it is possible to pass an additional inventory file (or even dynamic inventory script), using the hosts key. Ansible_ will automatically merge this inventory with the one generated by molecule.

---

> What is the purpose of this?

I pick up quite a bit data from the hcloud API about the instances and that is used by my infrastructure roles. This happens not only before the run but also during the run (`meta: refresh_inventory`) and it's especially necessary after fresh creation of a VM, because it didn't exist yet when the play started. An example for what this is used for is reconfiguring sshd to private net during hardening phase which makes the instance unreachable via the path that worked at the beginning of the playbook run.

All of this is unrelated to molecule tests, but it makes no sense to adjust the roles either. If there was a way to reduce the complexity in production code, I would already have happily done it. :)

> Are you saying the molecule servers are polluting your inventory?

No, that's not the problem. I try to describe it simplified and abstract:

1. hcloud project is empty (i.e., no instances exist)
1. execute `molecule test`
1. molecule creates instances as defined in `platforms` (e.g., named `foo` and `bar`) using the `molecule-hetznercloud` driver
    * now there are two instances in the hcloud project named `19fa9d8fc0b584b2b5e2704a69929e2d-foo` and `19fa9d8fc0b584b2b5e2704a69929e2d-bar`
1. this is when I refresh my inventory to pick up these instance and gather their metadata from hcloud API
1. now my roles under test get executed and expect these metadata vars be available
1. my role under test attempts to access such variable, a host_var of the host `foo` which it knows from the `platforms` attribute
    * this fails because in the inventory there is no host `foo` known that has these variables, because no hcloud instance named `foo` exists

At this point I applied my workaround that I mentioned in the first posting: In the dynamic inventory script I remove this prefix that was added by the driver, so the matching of host names works again and the expected variables are associated with the host as they should.

New info: I noticed the naming behavior changes completely when I use `--parallel` with `molecule test`. Then there is suddenly no such naming prefix anymore, but a random suffix. But that is currently not important for me because I have other means to parallelize the tests (in my CI) and it's currently good enough to have the per-role test package to be parallel, no need for scenarios to be parallelized as well.

> All the resources that molecule creates will have a label molecule={{ resource_namespace }}

I did not know that, thank you for the input! I guess I can make use of that in my inventory script to avoid the necessity to set a certain `resource_name`.

**Comment by jooola** [comment]
Ok, I start to better understand your situation.

I believe your best option is to continue using your workaround. Use a predefined molecule-namespace and reuse this prefix also in your inventory script (only adding the prefix when interacting with the Hetzner Cloud API).

The underlying problem is the difference between the molecule instance names and the Hetzner Cloud instance names, maybe having a flag to not use the namespace could be a good fix for you. With this feature, you will have to manage name uniqueness yourself.

If you have a better suggestion to improve this feel free to write your expectation here.

Let me know what you think.

**Comment by jooola** [comment]
Oh, and out of curiosity, what are the missing data that you are fetching from the API ? Would those be a good candidate for us to include them by default in the instance vars (molecule driver side) ?

**Comment by tumbl3w33d** [comment]
> If you have a better suggestion to improve this feel free to write your expectation here.

Thank you for your feedback and openness! Admittedly, at the moment I wouldn't have a better idea how to handle this in a generic way, but it is good enough for me and for sure good to have talked about it here for others with more exotic use cases like mine. I'm glad you showed up for maintaining this driver, by the way. I've been using the previous one for years (with the older molecule version) and needed to fork and patch both because nothing was happening here anymore. I was already afraid I would need to build all of this by myself. Thank you so much!

> Oh, and out of curiosity, what are the missing data that you are fetching from the API ?

I can show you a redacted example of my test run. At a very early stage (right when the instances are being created and even before private networks have been attached):
``` json
{
    "all": {
        "hosts": [
            "my-prefix-tumble-app-01",
            "my-prefix-tumble-app-02"
        ],
        "vars": {
            "ansible_user": "root"
        }
    },
    "_meta": {
        "hostvars": {
            "my-prefix-tumble-app-02": {
                "ansible_default_ipv4": {
                    "address": "<WAN IP>"
                },
                "hcloud_server_id": 44774548,
                "server_type": "cx21",
                "hcloud_image": "debian-11",
                "ansible_ssh_common_args": "-o ControlPath=\"~/.ssh/cp-%h-%p-%r\" -o ProxyCommand=\"ssh -W %h:%p -q jump-only.my.user@bastion.my.domain\"",
                "ansible_user": "my.user",
                "ansible_become": true,
                "ansible_host": "<WAN IP>"
            },
            "my-prefix-tumble-app-01": {
                "ansible_default_ipv4": {
                    "address": "<WAN IP>"
                },
                "hcloud_server_id": 44774549,
                "server_type": "cx21",
                "hcloud_image": "debian-11",
                "ansible_ssh_common_args": "-o ControlPath=\"~/.ssh/cp-%h-%p-%r\" -o ProxyCommand=\"ssh -W %h:%p -q jump-only.my.user@bastion.my.domain\"",
                "ansible_user": "my.user",
                "ansible_become": true,
                "ansible_host": "<WAN IP>"
            }
        }
    }
}
```
There's a story for each of them and a lot of it comes from the dynamic we have when bootstrapping fresh instances. In the past there was the common bastion approach so once sshd was hidden in a private net, we needed to detect that (I do a quick port scan in the inventory per instance) to figure out if the instance is publicly reachable yet or already hidden in the private net. Meanwhile we even improved that to have instances created in private networks right away (via user data of cloud init), but even then there's the dynamic that we start with a bootstrap port (2022) and at some point switch it to 22 and depending on that (port scan again) we route via bastion/VPN or not etc.

During the run, such instances changes settings and look like this
```
            "my-prefix-tumble-app-01": {
                "ansible_default_ipv4": {
                    "address": "<WAN IP>"
                },
                "hcloud_server_id": 44775035,
                "server_type": "cx21",
                "hcloud_image": "debian-11",
                "ansible_ssh_common_args": "-o ControlPath=\"~/.ssh/cp-%h-%p-%r\"",
                "private_network_ip": "10.0.0.2",
                "private_network_netmask": "255.255.255.255",
                "private_network_gateway": "10.0.0.1",
                "private_network_subnet_cidr": "10.0.0.0/16",
                "private_networks": [
                    {
                        "private_network_ip": "10.0.0.2",
                        "private_network_netmask": "255.255.255.255",
                        "private_network_gateway": "10.0.0.1",
                        "private_network_subnet_cidr": "10.0.0.0/16"
                    }
                ],
                "ansible_user": "root",
                "ansible_become": false,
                "ansible_host": "<WAN IP>"
            }
```
So I pick up all the private networks and by convention (actually historical reasons) one of them is determined to be the "primary" one to use. The way ansible connects also changes (in this case). You might wonder why I need to pick up some of these facts via hcloud API while I could just use those gathered by ansible, but for execution of bigger playbooks and setups it makes a big difference to have them ahead of time available without visiting all the involved hosts, especially when dealing with clusters or setting up per instance firewalls.
